dot/rpc/subscription/websocket_integration_test.go

```diff
@@ -18,13 +18,379 @@ import (
 )
 
-func TestWSConn_HandleConn(t *testing.T) {
+func TestWSConn_HandleConnParallel(t *testing.T) {
+	t.Parallel()
+
+	testCases := map[string]struct {
+		badRequest             bool
+		msg                    []byte
+		subscriptions          int
+		readErr, initErr       error
+		readErrMsg, initErrMsg string
+		storageAPIset          bool
+		reqID                  float64
+		params                 interface{}
+	}{
+		"test_StorageAPI_not_set": {
+			badRequest: true,
+			msg: []byte(`{"jsonrpc":"2.0",` +
+				`"error":{"code":null,"message":"error StorageAPI not set"},` +
+				`"id":1}` + "\n"),
+			subscriptions: 0,
+			initErr:       errStorageNotSet,
+			initErrMsg:    "error StorageAPI not set",
+			storageAPIset: false,
+			reqID:         1,
+			params:        nil,
+		},
+		"req_1_unexpected_type_nil": {
+			badRequest:    true,
+			msg:           nil,
+			subscriptions: 0,
+			initErr:       errUnexpectedType,
+			initErrMsg:    "unexpected type: <nil>, expected type []interface{}",
+			storageAPIset: true,
+			reqID:         1,
+			params:        nil,
+		},
+		"req_2": {
+			badRequest:    false,
+			msg:           []byte(`{"jsonrpc":"2.0","result":1,"id":2}` + "\n"),
+			subscriptions: 1,
+			initErr:       nil,
+			initErrMsg:    "",
+			storageAPIset: true,
+			reqID:         2,
+			params:        []interface{}{},
+		},
+		"req_3": {
+			badRequest:    false,
+			msg:           []byte(`{"jsonrpc":"2.0","result":1,"id":3}` + "\n"),
+			subscriptions: 1,
+			initErr:       nil,
+			initErrMsg:    "",
+			storageAPIset: true,
+			reqID:         3,
+			params:        []interface{}{"0x26aa"},
+		},
+		"req_1_unexpected_type_[]int": {
+			badRequest:    true,
+			msg:           nil,
+			subscriptions: 0,
+			initErr:       errUnexpectedType,
+			initErrMsg:    "unexpected type: []int, expected type string, []string, []interface{}",
+			storageAPIset: true,
+			reqID:         1,
+			params:        []interface{}{[]int{123}},
+		},
+		"req_1_unexpected_type_int": {
+			badRequest:    true,
+			msg:           nil,
+			subscriptions: 0,
+			initErr:       errUnexpectedType,
+			initErrMsg:    "unexpected type: int, expected type string, []string, []interface{}",
+			storageAPIset: true,
+			reqID:         1,
+			params:        []interface{}{123},
+		},
+	}
+	for name, testCase := range testCases {
+		testCase := testCase
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			ctrl := gomock.NewController(t)
+
+			wsconn, c, cancel := setupWSConn(t)
+			wsconn.Subscriptions = make(map[uint32]Listener)
+			if testCase.storageAPIset {
+				wsconn.StorageAPI = modules.NewMockAnyStorageAPI(ctrl)
+			}
+			defer cancel()
+			go wsconn.HandleConn()
+			time.Sleep(time.Second * 2)
+			res, initErr := wsconn.initStorageChangeListener(testCase.reqID, testCase.params)
+			if testCase.badRequest {
+				require.Nil(t, res)
+			} else {
+				require.NotNil(t, res)
+			}
+			require.Equal(t, len(wsconn.Subscriptions), testCase.subscriptions)
```

Comment:
Can you use `require.Len` instead of `require.Equal`?

```suggestion
			require.Len(t, wsconn.Subscriptions, testCase.subscriptions)
```